Agents/TravellingSalesmanAgent/travellingsalesmanagent/src/main/java/uk/ac/cam/cares/jps/agent/travellingsalesmanagent/TSPRouteGenerator.java

```diff
@@ -0,0 +1,120 @@
+package uk.ac.cam.cares.jps.agent.travellingsalesmanagent;
+
+
+import com.cmclinnovations.stack.clients.geoserver.GeoServerClient;
+import com.cmclinnovations.stack.clients.geoserver.GeoServerVectorSettings;
+import com.cmclinnovations.stack.clients.geoserver.UpdatedGSVirtualTableEncoder;
+
+public class TSPRouteGenerator {
+
+
+    /**
+     * Generate TSP route layer
+     * @param geoServerClient
+     * @param workspaceName
+     * @param schema
+     * @param dbName
+     * @param LayerName
+     * @param cost_table
+     */
+    public void generateTSPLayer(GeoServerClient geoServerClient, String workspaceName, String schema, String dbName, String LayerName, String cost_table){
+
+        String tspLayer = "WITH tsp AS (\n" +
+                "    SELECT *\n" +
+                "    FROM pgr_TSP(\n" +
+                "        $$SELECT * FROM pgr_dijkstraCostMatrix(\n" +
+                "            '"+cost_table+"',\n" +
+                "            (\n" +
+                "                SELECT ARRAY_APPEND(array_agg(id), (%target%))\n" +
+                "                FROM routing_ways_vertices_pgr\n" +
+                "                WHERE id IN (SELECT DISTINCT nearest_node FROM poi_tsp_nearest_node, flood_polygon_single_10cm WHERE ST_Intersects(poi_tsp_nearest_node.geom, flood_polygon_single_10cm.geom) OR ST_DISTANCE (poi_tsp_nearest_node.geom, flood_polygon_single_10cm.geom) < 0.005)\n" +
+                "            ),\n" +
+                "            false\n" +
+                "        )$$, \n" +
+                "        (%target%),"+
+                "        (%target%)"+
+                "    )\n" +
+                "),\n" +
+                "tsp_route AS (\n" +
+                "SELECT routing_ways.the_geom\n" +
+                "FROM (\n" +
+                "    SELECT ROW_NUMBER() OVER() as seq, tsp.node\n" +
+                "    FROM tsp\n" +
+                ") n1\n" +
+                "JOIN (\n" +
+                "    SELECT ROW_NUMBER() OVER() as seq, tsp.node\n" +
+                "    FROM tsp\n" +
+                ") n2 ON n1.seq + 1 = n2.seq\n" +
+                "JOIN pgr_dijkstra(\n" +
+                "    '"+cost_table+"',\n" +
+                "    n1.node,\n" +
+                "    n2.node, false\n" +
+                ") AS di ON true\n" +
+                "JOIN routing_ways ON di.edge = routing_ways.gid\n" +
+                "ORDER BY n1.seq\n" +
+                ")\n" +
+                "SELECT tsp_route.the_geom as geom\n" +
+                "FROM tsp_route ";
```

Comment:
Think in general it would be better to put the SQL queries in a different file in the resources folder but probably no need to change it here

Reply:
That is a great suggestion and good point. The SQL takes input -"cost_table" as parameter, maybe it would be good to use a SQL builder for it to look cleaner instead of using resources folder. Will make a note and refractor it on the next update! Thanks!